src/Dependencies/Collections/ImmutableSegmentedDictionary`2.cs

```diff
@@ -0,0 +1,360 @@
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics.CodeAnalysis;
+
+namespace Microsoft.CodeAnalysis.Collections
+{
+    /// <summary>
+    /// Represents a segmented dictionary that is immutable; meaning it cannot be changed once it is created.
+    /// </summary>
+    /// <remarks>
+    /// <para>There are different scenarios best for <see cref="ImmutableSegmentedDictionary{TKey, TValue}"/> and others
+    /// best for <see cref="ImmutableDictionary{TKey, TValue}"/>.</para>
+    ///
+    /// <para>In general, <see cref="ImmutableSegmentedDictionary{TKey, TValue}"/> is applicable in scenarios most like
+    /// the scenarios where <see cref="ImmutableArray{T}"/> is applicable, and
+    /// <see cref="ImmutableDictionary{TKey, TValue}"/> is applicable in scenarios most like the scenarios where
+    /// <see cref="ImmutableList{T}"/> is applicable.</para>
+    ///
+    /// <para>The following table summarizes the performance characteristics of
+    /// <see cref="ImmutableSegmentedDictionary{TKey, TValue}"/>:</para>
+    /// 
+    /// <list type="table">
+    ///   <item>
+    ///     <description>Operation</description>
+    ///     <description><see cref="ImmutableSegmentedDictionary{TKey, TValue}"/> Complexity</description>
+    ///     <description><see cref="ImmutableDictionary{TKey, TValue}"/> Complexity</description>
+    ///     <description>Comments</description>
+    ///   </item>
+    ///   <item>
+    ///     <description>Item</description>
+    ///     <description>O(1)</description>
+    ///     <description>O(log n)</description>
+    ///     <description>Directly index into the underlying segmented dictionary</description>
+    ///   </item>
+    ///   <item>
+    ///     <description>Add()</description>
+    ///     <description>O(n)</description>
+    ///     <description>O(log n)</description>
+    ///     <description>Requires creating a new segmented dictionary</description>
+    ///   </item>
+    /// </list>
+    /// 
+    /// <para>This type is backed by segmented arrays to avoid using the Large Object Heap without impacting algorithmic
+    /// complexity.</para>
+    /// </remarks>
+    /// <typeparam name="TKey">The type of the keys in the dictionary.</typeparam>
+    /// <typeparam name="TValue">The type of the values in the dictionary.</typeparam>
+    /// <devremarks>
+    /// <para>This type has a documented contract of being exactly one reference-type field in size. Our own
+    /// <see cref="RoslynImmutableInterlocked"/> class depends on it, as well as others externally.</para>
+    ///
+    /// <para><strong>IMPORTANT NOTICE FOR MAINTAINERS AND REVIEWERS:</strong></para>
+    ///
+    /// <para>This type should be thread-safe. As a struct, it cannot protect its own fields from being changed from one
```

Comment:
>This [](start = 14, length = 4)

Would it make sense to move implementation of each instance member to a static nested class to make this more obvious?

```Foo(...) => FooImpl(this, ...);```

Reply:
In theory it could, but there are two mitigating factors:

1. The current code follows the pattern used for `ImmutableArray<T>`
2. The API surface for this type isn't expected to change significantly or frequently

IMO the split would add a relatively large amount of code for minimal long-term gains.

Reply:
Sounds good.